src/common/get-card-selection-view-data.ts

```diff
@@ -1,65 +1,147 @@
+) => CardSelectionViewData;
+
+export const getCardSelectionViewData: GetCardSelectionViewData = (
+    cardSelectionStoreData: CardSelectionStoreData,
+    unifiedScanResultStoreData: UnifiedScanResultStoreData,
+    getUnavailableHighlightStatus: GetUnavailableHighlightStatus,
 ): CardSelectionViewData => {
     const viewData = getEmptyViewData();
 
-    if (!storeData) {
+    if (
+        isEmpty(cardSelectionStoreData) ||
+        isEmpty(unifiedScanResultStoreData) ||
+        unifiedScanResultStoreData.results == null
+    ) {
         return viewData;
     }
 
-    viewData.visualHelperEnabled = storeData.visualHelperEnabled || false;
+    viewData.visualHelperEnabled = cardSelectionStoreData.visualHelperEnabled || false;
+    viewData.expandedRuleIds = getRuleIdsOfExpandedRules(cardSelectionStoreData.rules);
 
-    viewData.expandedRuleIds = getRuleIdsOfExpandedRules(storeData.rules);
+    const allResultUids = getAllResultUids(cardSelectionStoreData.rules);
+    const resultsWithUnavailableHighlightStatus = getResultsWithUnavailableHighlightStatus(
+        unifiedScanResultStoreData,
+        getUnavailableHighlightStatus,
+    );
+
+    allResultUids.forEach(resultUid => {
+        viewData.resultsHighlightStatus[resultUid] = 'hidden';
+    });
 
-    if (!storeData.visualHelperEnabled) {
-        // no selected cards; no highlighted instances
+    if (!cardSelectionStoreData.visualHelperEnabled) {
+        // no selected cards; no highlighted instances and unavailable instances shown as such.
+        viewData.resultsHighlightStatus = {
+            ...viewData.resultsHighlightStatus,
+            ...resultsWithUnavailableHighlightStatus,
+        };
         return viewData;
     }
 
+    // if no rules are expanded, highlight all results except for those unavailable.
     if (viewData.expandedRuleIds.length === 0) {
-        viewData.highlightedResultUids = getAllResultUids(storeData.rules);
+        forOwn(viewData.resultsHighlightStatus, (_, resultUid) => {
+            viewData.resultsHighlightStatus[resultUid] = 'visible';
+        });
+        viewData.resultsHighlightStatus = {
+            ...viewData.resultsHighlightStatus,
+            ...resultsWithUnavailableHighlightStatus,
+        };
         return viewData;
     }
 
+    // if there are any cards selected, make sure they are highlighted if not unavailable.
     viewData.selectedResultUids = getOnlyResultUidsFromSelectedCards(
-        storeData.rules,
+        cardSelectionStoreData.rules,
         viewData.expandedRuleIds,
     );
 
-    viewData.highlightedResultUids = viewData.selectedResultUids.length
-        ? viewData.selectedResultUids
-        : getAllResultUidsFromRuleIdArray(storeData.rules, viewData.expandedRuleIds);
+    if (viewData.selectedResultUids.length) {
+        viewData.selectedResultUids.forEach(resultUid => {
+            viewData.resultsHighlightStatus[resultUid] = 'visible';
+        });
+        viewData.resultsHighlightStatus = {
```

Comment:
I agree that the `resultsHighlightStatus` object is a nicer representation for consumers of the view data, but I think the implementation of this method has become too complex (88 lines). I tried out some options for simplifying and I think my favorite was with separating the concerns of identifying the visible UIDs vs consolidating the "hidden, visible, unavailable" lists into the combined highlight status object. Consider:

```typescript
export const getCardSelectionViewData: GetCardSelectionViewData = (
    cardSelectionStoreData: CardSelectionStoreData,
    unifiedScanResultStoreData: UnifiedScanResultStoreData,
    getUnavailableHighlightStatus: GetUnavailableHighlightStatus,
): CardSelectionViewData => {
    const allRules = cardSelectionStoreData.rules;
    const viewData = getEmptyViewData();

    if (isEmpty(cardSelectionStoreData) || unifiedScanResultStoreData.results == null) {
        return viewData;
    }

    viewData.visualHelperEnabled = cardSelectionStoreData.visualHelperEnabled || false;
    viewData.expandedRuleIds = getExpandedRuleIds(allRules);
    viewData.selectedResultUids = getSelectedResultUids(allRules, viewData.expandedRuleIds);

    const allResultUids = getAllResultUids(allRules);
    const unavailableResultUids = getUnavailableResultUids(
        unifiedScanResultStoreData,
        getUnavailableHighlightStatus,
    );

    let visibleResultUids: string[];
    if (!cardSelectionStoreData.visualHelperEnabled) {
        visibleResultUids = [];
    } else if (viewData.selectedResultUids.length > 0) {
        visibleResultUids = viewData.selectedResultUids;
    } else {
        visibleResultUids = getAllResultUidsForRuleIds(allRules, viewData.expandedRuleIds);
    }

    viewData.resultsHighlightStatus = getHighlightStatusByResultUid(
        allResultUids,
        visibleResultUids,
        unavailableResultUids,
    );

    return viewData;
};

// ...

function getUnavailableResultUids(
    unifiedScanResultStoreData: UnifiedScanResultStoreData,
    getUnavailableHighlightStatus: GetUnavailableHighlightStatus,
): string[] {
    const { platformInfo } = unifiedScanResultStoreData;

    return unifiedScanResultStoreData.results
        .filter(
            result =>
                result.status === 'fail' &&
                getUnavailableHighlightStatus(result, platformInfo) === 'unavailable',
        )
        .map(result => result.uid);
}

function getHighlightStatusByResultUid(
    allResultUids: string[],
    visibleResultUids: string[],
    unavailableResultUids: string[],
): ResultsHighlightStatus {
    const result = {};
    for (const resultUid of allResultUids) {
        result[resultUid] = 'hidden';
    }
    for (const resultUid of visibleResultUids) {
        result[resultUid] = 'visible';
    }
    for (const resultUid of unavailableResultUids) {
        result[resultUid] = 'unavailable';
    }
    return result;
}
```


```

Reply:
Agreed.

Thanks for the code suggestions as well; let me look into making this a bit simpler/clearer.